File: cs/Parking/ParkingLot.cs
using System;
using System.Collections.Generic;

namespace parking
{
    public class ParkingLot
    {
        public ParkingLotID ID { get; private set; }
        private int _capacity;
        private readonly IDictionary<Ticket, Car> _parkedCars;

        public ParkingLot(int totalPosition)
        {
            ID = new ParkingLotID(Guid.NewGuid().ToString());
            _capacity = totalPosition;
            _parkedCars = new Dictionary<Ticket, Car>();
        }

        public Ticket Park(Car car)
        {
            if (_capacity <= 0)
            {
                throw new Exception("capacity is not enough.");
            }

            var ticket = new Ticket(car.PlateNumber, ID);
            _parkedCars.Add(ticket, car);
            _capacity--;
            return ticket;
        }

        public Car Take(Ticket ticket)
        {
            if (!ticket.IsValid || !_parkedCars.ContainsKey(ticket))
            {
                throw new Exception("Ticket is invalided.");
            }

            var car = _parkedCars[ticket];
            ticket.Invalidate();
            _parkedCars.Remove(ticket);
            _capacity++;
            return car;
        }
    }

    public class ParkingLotID : ValueObject<string>
    {
        public ParkingLotID(string parkingLotID) : base(parkingLotID) { }
    }
}

File: ts/src/parking/domain/parkingLot/parking/parkingLot.ts
/** @format */

import {Car} from './car'
import {Ticket} from './ticket'
import {ValueObject} from '../../valueObject'

export class ParkingLot {
    private _availableSpaces: number
    private parkedCars: Map<Ticket, Car>
    public readonly id: ParkingLotId

    constructor(id: string, totalSpaces: number) {
        this._availableSpaces = totalSpaces
        this.parkedCars = new Map<Ticket, Car>()
        this.id = new ParkingLotId(id)
    }

    public get availableSpaces() {
        return this._availableSpaces
    }

    public get hasAvailableSpaces() {
        return this._availableSpaces > 0
    }

    public park(car: Car) {
        if (this._availableSpaces <= 0) {
            throw new Error('capacity is not enough.')
        }

        const ticket = new Ticket(car.plant, this.id)
        this.parkedCars.set(ticket, car)
        this._availableSpaces--
        return ticket
    }

    public take(ticket: Ticket) {
        if (!ticket.isValid || !this.parkedCars.has(ticket)) {
            throw new Error('Ticket is invalided.')
        }

        const car = this.parkedCars.get(ticket)
        ticket.invalidate()
        this.parkedCars.delete(ticket)
        this._availableSpaces++
        return car
    }
}

export class ParkingLotId extends ValueObject<String> {
    constructor(parkingLotID: String) {
        super(parkingLotID)
    }
}
